spec.html

```diff
@@ -7872,6 +7877,149 @@ <h1>Forward Progress</h1>
       <p>This, along with the liveness guarantee in the memory model, ensures that all ~SeqCst~ writes eventually become observable to all agents.</p>
     </emu-note>
   </emu-clause>
+
+  <emu-clause id="sec-weakref-processing-model">
+    <h1>Processing model of WeakRef and FinalizationRegistry objects</h1>
+    <emu-clause id="sec-weakref-invariants">
+      <h1>Objectives</h1>
+
+      <p> This specification does not make any guarantees that any object will be garbage collected. Objects which are not live may be released after long periods of time, or never at all. For this reason, this specification uses the term "may" when describing behaviour triggered by garbage collection.</p>
+
+      <p>The semantics of WeakRef and FinalizationRegistry objects is based on two operations which happen at particular points in time: </p>
+
+      <ul>
+        <li>
+          When `WeakRef.prototype.deref` is called, the referent (if *undefined* is not returned) is kept alive so that subsequent, synchronous accesses also return the object. This list is reset when synchronous work is done using the ClearKeptObjects abstract operation.
+        </li>
+
+        <li>
+          When an object which is registered with a FinalizationRegistry becomes unreachable, a call of the FinalizationRegistry's cleanup callback may eventually be made, after synchronous ECMAScript execution completes.  The FinalizationRegistry cleanup is performed with the CleanupFinalizationRegistry abstract operation.
+        </li>
+      </ul>
+
+      <p>Neither of these actions (ClearKeptObjects or CleanupFinalizationRegistry) may interrupt synchronous ECMAScript execution. Because host environments may assemble longer, synchronous ECMAScript execution runs, this specification defers the scheduling of ClearKeptObjects and CleanupFinalizationRegistry to the host environment.</p>
+
+      <p>Some ECMAScript implementations include garbage collector implementations which run in the background, including when ECMAScript is idle. Letting the host environment schedule CleanupFinalizationRegistry allows it to resume ECMAScript execution in order to run finalizer work, which may free up held values, reducing overall memory usage.</p>
+    </emu-clause>
+
+    <emu-clause id="sec-liveness">
+      <h1>Liveness</h1>
+
+      <p>For some set of objects _S_, a <dfn>hypothetical WeakRef-oblivious</dfn> execution with respect to _S_ is an execution whereby `WeakRef.prototype.deref` being called on a WeakRef whose referent is an element of _S_ always returns *undefined*.</p>
```

Comment:
The intention is that it's cross-Realm but WeakRefs are not cross-agent. Agents, outside of SABs, have disjoint object closures in practice (though as you know this isn't exactly formally spelled out).

You are correct though that the intention is that no matter which particular Realm's `WeakRef.prototype.deref` is called within an agent, this constraint applies. It's kind of unfortunate if that's the only reason we need to lift out the steps of `WeakRef.prototype.deref` into an AO, but the alternative is to say something like "the algorithm steps of `WeakRef.prototype.deref` being evaluated", which is worse. If we had an AO it'd also link straightforwardly.

All in all I agree refactoring into an AO is a good fix here.

Reply:
It's possible we could find a way of phrasing this so that it was talking about any realm's % WeakRef.prototype.deref%, without needing the new AO, but I don't have a phrasing I'm happy with. @codehag was going to try to think of something.

Reply:
I suspect there's other places in the spec where an intrinsic is discussed, and is intended to refer to *every* realm's copy of that intrinsic.

Reply:
Just audited the spec; you're correct that there are no others.

Reply:
Proposed wording (with help from @littledan): 

For some set of objects S, a <dfn>hypothetical WeakRef-oblivious</dfn> execution with respect to S is an execution whereby, for every Realm r, r.[[Intrinsics]].[[%WeakRef.prototype.deref%]] being called on a WeakRef whose referent is an element of S always returns undefined.

Reply:
Also, it may be easier to do this as an AO in the end

Reply:
AO as @bakkot and @syg suggested seems simplest, as it removes the need to think about Realms at all.

Reply:
I have drafted a version of the AO. It's the first time I do it so likely needs an iteration.